General queries/Baseline Comparison.txt

```diff
@@ -0,0 +1,239 @@
+//   4. Image Loads (disabled by default due to volume, enable at your own risk, additional data has initiating processes)
+//   5. Logon (derived from LogonEvents for the unique users logged on, additional data has logon types)
+//   6. Network communication (grouped by 2nd level-domain, ie 'microsoft.com' in 'www.microsoft.com' and 'web.microsoft.com', additional data has the full list of URLs)
+//   7. Process creation (additional data has the full paths of the files)
+//   8. Powershell Commands (grouped by the cmdlet that was ran, additional data has the processes that ran the cmdlet)
+//   9. Reigstry Events (disabled by default due to volume, grouped by the registry key, additional data has the value data)
+//  10. Raw IP Connection Events (additional data has the initiating processes)
+//
+// List of "known good" hosts - populate with your baseline, must be FQDNs
+let GoodHosts=pack_array('supposedlygoodhost.mydomain','ithinkitsgoodserver.mydomain');
+// List of suspected bad hosts - populate with bad machines, must be FQDNs
+let SuspectedBadHosts=pack_array('compromisedhost.mydomain', 'lateralmovementhost.mydomain');
+// How far back should the baseline be built from?
+let GoodTimeRange=30d;
+// How far back should the bad machines be looked at?
+let SuspectedBadTimeRange=30d;
+// Comment return sets that you do not want returned, by default file creation and image loads and registry events are disabled
+let ReturnSets=pack_array(
+    'Alert',
+    'Connected Networks',
+//    'File Creation'
+//    'Image Loads',
+    'Logon',
+    'Network Communication',
+    'Process Creation',
+    'PowerShell Command',
+//    'Registry Event'
+    'Raw IP Communication'
+);
+// -------------End of variables, changing below this line will change query logic----------
+// Function to get a list of machine IDs given a list of computer names
+let GetMachineId=(InComputerName: dynamic) {
+    MachineInfo
+    | where ComputerName in~ (InComputerName)
+    | summarize makeset(MachineId)
+};
+// Function to get network communications given a list of computer names and how far back to look
+let GetNetworkEvents=(InMachineId: dynamic, LeftEventTime: datetime) {
+    NetworkCommunicationEvents
+    | where "Network Communication" in (ReturnSets)
+    | where EventTime > LeftEventTime
+    | where MachineId in~ (InMachineId)
+    | summarize EventTime=max(EventTime), count() by RemoteUrl, MachineId
+    | extend UrlSplit=split(RemoteUrl, ".") // Split the levels of the URL
+    // If there is only one level (for an internal communication that uses your DNS search suffix), then only use that level
+    // Otherwise combine the top two levels and use those as the URLRoot
+    | extend UrlRoot=iff(UrlSplit[-2] == "", UrlSplit[0], strcat(tostring(UrlSplit[-2]), ".", tostring(UrlSplit[-1])))
+    | summarize EventTime=max(EventTime), Count=sum(count_), AdditionalData=makeset(RemoteUrl) by UrlRoot, MachineId
```

Comment:
better limit the makeset, e.g. makeset(RemoteUrl, 5)

Reply:
Updated in latest commit.